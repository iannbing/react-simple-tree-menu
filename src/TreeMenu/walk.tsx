import isEmpty from 'is-empty';
import memoize from 'fast-memoize';

export interface TreeNodeObject {
  [name: string]: TreeNode;
}

interface LocaleFunctionProps {
  label: string;
  [name: string]: any;
}

interface MatchSearchFunctionProps extends LocaleFunctionProps {
  searchTerm: string;
}

export interface TreeNode extends LocaleFunctionProps {
  index: number;
  nodes?: TreeNodeObject;
}

export interface TreeNodeInArray extends LocaleFunctionProps {
  key: string;
  nodes?: TreeNodeInArray[];
}

export type LocaleFunction = (localeFunctionProps: LocaleFunctionProps) => string;
export type MatchSearchFunction = (
  matchSearchFunctionProps: MatchSearchFunctionProps
) => boolean;

type Data = TreeNodeObject | TreeNodeInArray[];
interface WalkProps {
  data: Data | undefined;
  parent?: string;
  level?: number;
  openNodes: string[];
  searchTerm: string;
  keySeparator?: string;
  locale?: LocaleFunction;
  matchSearch?: MatchSearchFunction;
}

interface BranchProps {
  parent: string;
  level: number;
  openNodes: string[];
  searchTerm: string;
  node: TreeNode | TreeNodeInArray;
  nodeName: string;
  index?: number;
  keySeparator?: string;
  locale?: LocaleFunction;
  matchSearch?: MatchSearchFunction;
}

export interface Item {
  hasNodes: boolean;
  isOpen: boolean;
  level: number;
  key: string;
  label: string;
  [name: string]: any;
}

const validateData = (data: Data | undefined): boolean => !!data && !isEmpty(data);
const getValidatedData = (data: Data | undefined) =>
  validateData(data) ? (data as Data) : [];

const walk = ({ data, ...props }: WalkProps): Item[] => {
  const validatedData = getValidatedData(data);

  const propsWithDefaultValues = { parent: '', level: 0, keySeparator: '/', ...props };
  const handleArray = (dataAsArray: TreeNodeInArray[]) =>
    dataAsArray.reduce((all: Item[], node: TreeNodeInArray, index) => {
      const branchProps = { node, index, nodeName: node.key, ...propsWithDefaultValues };
      const branch = generateBranch(branchProps);
      return [...all, ...branch];
    }, []);

  const handleObject = (dataAsObject: TreeNodeObject) =>
    Object.entries(dataAsObject)
      .sort((a, b) => a[1].index - b[1].index) // sorted by index
      .reduce((all: Item[], [nodeName, node]: [string, TreeNode]) => {
        const branchProps = { node, nodeName, ...propsWithDefaultValues };
        const branch = generateBranch(branchProps);
        return [...all, ...branch];
      }, []);

  return Array.isArray(validatedData)
    ? handleArray(validatedData)
    : handleObject(validatedData);
};

const defaultMatchSearch = ({ label, searchTerm }: MatchSearchFunctionProps) => {
  const processString = (text: string): string => text.trim().toLowerCase();
  return processString(label).includes(processString(searchTerm));
};

const defaultLocale = ({ label }: LocaleFunctionProps): string => label;

const generateBranch = ({
  node,
  nodeName,
  matchSearch = defaultMatchSearch,
  locale = defaultLocale,
  ...props
}: BranchProps): Item[] => {
  const { parent, level, openNodes, searchTerm } = props;

  const { nodes, label: rawLabel = 'unknown', ...nodeProps } = node;
  const key = [parent, nodeName].filter(x => x).join(props.keySeparator);
  const hasNodes = validateData(nodes);
  const isOpen = hasNodes && (openNodes.includes(key) || !!searchTerm);

  const label = locale({ label: rawLabel, ...nodeProps });
  const isVisible = !searchTerm || matchSearch({ label, searchTerm, ...nodeProps });
  const currentItem = { ...props, ...nodeProps, label, hasNodes, isOpen, key };

  const data = getValidatedData(nodes);
  const nextLevelItems = isOpen
    ? walk({ data, locale, matchSearch, ...props, parent: key, level: level + 1 })
    : [];

  return isVisible ? [currentItem, ...nextLevelItems] : nextLevelItems;
};

export const fastWalk = memoize(walk);
export const slowWalk = walk;
