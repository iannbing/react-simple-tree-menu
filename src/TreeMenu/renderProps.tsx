import React, { ReactNode } from 'react';
import classNames from 'classnames';

import { Item } from './walk';

const DEFAULT_PADDING = 0.75;
const ICON_SIZE = 2;
const LEVEL_SPACE = 1.75;
const ToggleIcon = ({
  on,
  openedIcon,
  closedIcon,
}: {
  on: boolean;
  openedIcon: ReactNode;
  closedIcon: ReactNode;
}) => (
  <div role="img" aria-label="Toggle" className="toggle-icon-symbol">
    {on ? openedIcon : closedIcon}
  </div>
);

export interface TreeMenuItem extends Item {
  active?: boolean;
  onClick: (event: React.MouseEvent<HTMLLIElement>) => void;
  toggleNode?: () => void;
}

export type TreeMenuChildren = (props: {
  search?: (term: string) => void;
  searchTerm?: string;
  items: TreeMenuItem[];
  resetOpenNodes?: (openNodes?: string[]) => void;
}) => JSX.Element;

export const ItemComponent: React.FunctionComponent<TreeMenuItem> = ({
  hasNodes = false,
  isOpen = false,
  level = 0,
  onClick,
  toggleNode,
  active,
  focused,
  openedIcon = '-',
  closedIcon = '+',
  label = 'unknown',
  style = {},
}) => (
  <li
    className={classNames(
      'tree-item',
      { 'tree-item--active': active },
      { 'tree-item--focused': focused }
    )}
    style={{
      paddingLeft: `${DEFAULT_PADDING +
        ICON_SIZE * (hasNodes ? 0 : 1) +
        level * LEVEL_SPACE}rem`,
      ...style,
    }}
    role="button"
    aria-pressed={active}
    onClick={onClick}
  >
    {hasNodes && (
      <div
        className="toggle-icon"
        onClick={e => {
          hasNodes && toggleNode && toggleNode();
          e.stopPropagation();
        }}
      >
        <ToggleIcon on={isOpen} openedIcon={openedIcon} closedIcon={closedIcon} />
      </div>
    )}
    {label}
  </li>
);

export const defaultChildren: TreeMenuChildren = ({ search, items }) => {
  const onSearch = (e: React.ChangeEvent<HTMLInputElement>) => {
    const { value } = e.target;
    search && search(value);
  };
  return (
    <>
      {search && (
        <input
          className="tree-search"
          aria-label="Type and search"
          type="search"
          placeholder="Type and search"
          onChange={onSearch}
        />
      )}
      <ul className="tree-item-group">
        {items.map(({ key, ...props }) => (
          <ItemComponent key={key} {...props}></ItemComponent>
        ))}
      </ul>
    </>
  );
};
