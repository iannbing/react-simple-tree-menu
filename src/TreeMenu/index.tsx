import React from 'react';
import debounce from 'tiny-debounce';

import  {
  fastWalk,
  slowWalk,
  TreeNode,
  Item,
  TreeNodeInArray,
  LocaleFunction,
  MatchSearchFunction,
} from './walk';
import { defaultChildren, TreeMenuChildren, TreeMenuItem } from './renderProps';
import KeyDown from '../KeyDown';

export type TreeMenuProps = {
  data: { [name: string]: TreeNode } | TreeNodeInArray[];
  activeKey?: string;
  focusKey?: string;
  keySeparator: string;
  initialActiveKey?: string;
  initialFocusKey?: string;
  initialOpenNodes?: string[];
  openNodes?: string[];
  resetOpenNodesOnDataUpdate?: boolean;
  hasSearch?: boolean;
  cacheSearch?: boolean;
  onClickItem?: (props: Item) => void;
  debounceTime?: number;
  children?: TreeMenuChildren;
  locale?: LocaleFunction;
  matchSearch?: MatchSearchFunction;
  disableKeyboard?: boolean;
};

type TreeMenuState = {
  openNodes: string[];
  searchTerm: string;
  activeKey: string;
  focusKey: string;
};

const defaultOnClick = (props: Item) => console.log(props); // eslint-disable-line no-console

class TreeMenu extends React.Component<TreeMenuProps, TreeMenuState> {
  static defaultProps: TreeMenuProps = {
    data: {},
    onClickItem: defaultOnClick,
    debounceTime: 125,
    children: defaultChildren,
    hasSearch: true,
    cacheSearch:true,
    resetOpenNodesOnDataUpdate: false,
    disableKeyboard: false,
    keySeparator: '/',
  };

  state: TreeMenuState = {
    openNodes: this.props.initialOpenNodes || [],
    searchTerm: '',
    activeKey: this.props.initialActiveKey || '',
    focusKey: this.props.initialFocusKey || '',
  };

  componentDidUpdate(prevProps: TreeMenuProps) {
    const { data, initialOpenNodes, resetOpenNodesOnDataUpdate } = this.props;
    if (prevProps.data !== data && resetOpenNodesOnDataUpdate && initialOpenNodes) {
      this.setState({ openNodes: initialOpenNodes });
    }
  }

  resetOpenNodes = (newOpenNodes?: string[], activeKey?: string, focusKey?: string) => {
    const { initialOpenNodes } = this.props;
    const openNodes =
      (Array.isArray(newOpenNodes) && newOpenNodes) || initialOpenNodes || [];
    this.setState({ openNodes, searchTerm: '', activeKey: activeKey || '', focusKey: focusKey || activeKey || '' });
  };

  search = (value: string) => {
    const { debounceTime } = this.props;
    const search = debounce(
      (searchTerm: string) => this.setState({ searchTerm }),
      debounceTime
    );
    search(value);
  };

  toggleNode = (node: string) => {
    if (!this.props.openNodes) {
      const { openNodes } = this.state;
      const newOpenNodes = openNodes.includes(node)
        ? openNodes.filter(openNode => openNode !== node)
        : [...openNodes, node];
      this.setState({ openNodes: newOpenNodes });
    }
  };

  generateItems = (): TreeMenuItem[] => {
    const { data, onClickItem, locale, matchSearch, keySeparator } = this.props;
    const { searchTerm } = this.state;
    const openNodes = this.props.openNodes || this.state.openNodes;
    const activeKey = this.props.activeKey || this.state.activeKey;
    const focusKey = this.props.focusKey || this.state.focusKey;
    const defaultSearch = this.props.cacheSearch ? fastWalk : slowWalk;
    const items: Item[] = data
      ? defaultSearch({ data, openNodes, searchTerm, locale, matchSearch, keySeparator })
      : [];

    return items.map(item => {
      const focused = item.key === focusKey;
      const active = item.key === activeKey;
      const onClick = () => {
        const newActiveKey = this.props.activeKey || item.key;
        this.setState({ activeKey: newActiveKey, focusKey: newActiveKey });
        onClickItem && onClickItem(item);
      };

      const toggleNode = item.hasNodes ? () => this.toggleNode(item.key) : undefined;
      return { ...item, focused, active, onClick, toggleNode };
    });
  };

  getKeyDownProps = (items: TreeMenuItem[]) => {
    const { onClickItem } = this.props;
    const { focusKey, activeKey, searchTerm } = this.state;
    
    const focusIndex = items.findIndex(item => item.key === (focusKey || activeKey));
    const getFocusKey = (item: TreeMenuItem) => {
      const keyArray = item.key.split(this.props.keySeparator);

      return keyArray.length > 1
        ? keyArray.slice(0, keyArray.length - 1).join(this.props.keySeparator)
        : item.key;
    };

    return {
      up: () => {
        this.setState(({ focusKey }) => ({
          focusKey: focusIndex > 0 ? items[focusIndex - 1].key : focusKey,
        }));
      },
      down: () => {
        this.setState(({ focusKey }) => ({
          focusKey: focusIndex < items.length - 1 ? items[focusIndex + 1].key : focusKey,
        }));
      },
      left: () => {
        const item = items[focusIndex];
        if (item) {
          this.setState(({ openNodes, ...rest }) => {
            const newOpenNodes = openNodes.filter(node => node !== item.key);
            return item.isOpen
              ? { ...rest, openNodes: newOpenNodes, focusKey: item.key }
              : { ...rest, focusKey: getFocusKey(item) };
          });
        }
      },
      right: () => {
        const { hasNodes, key } = items[focusIndex];
        if (hasNodes)
          this.setState(({ openNodes }) => ({ openNodes: [...openNodes, key] }));
      },
      enter: () => {
        this.setState(({ focusKey }) => ({ activeKey: focusKey }));
        onClickItem && onClickItem(items[focusIndex]);
      },
    };
  };

  render() {
    const { children, hasSearch, disableKeyboard } = this.props;
    const { searchTerm } = this.state;

    const search = this.search;
    const items = this.generateItems();
    const resetOpenNodes = this.resetOpenNodes;
    const render = children || defaultChildren;

    const renderProps = hasSearch
      ? {
          search,
          resetOpenNodes,
          items,
          searchTerm,
        }
      : { items, resetOpenNodes };

    return disableKeyboard ? (
      render(renderProps)
    ) : (
      <KeyDown {...this.getKeyDownProps(items)}>{render(renderProps)}</KeyDown>
    );
  }
}

export default TreeMenu;
