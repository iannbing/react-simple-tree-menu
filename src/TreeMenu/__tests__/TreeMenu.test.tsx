import React from 'react';
import { shallow, mount } from 'enzyme';

import TreeMenu from '../index';

const mockData = {
  atd: {
    label: 'ATS Guide',
    key: 'ats',
    index: 1, // ATS Guide should be after Release Notes
  },
  releasenotes: {
    label: 'Release Notes',
    key: 'releasenotes',
    index: 0, // Release Notes should be first
    nodes: {
      'desktop-modeler': {
        label: 'Desktop Modeler',
        key: 'releasenotes/desktop-modeler',
        index: 0,
        nodes: {
          7: {
            label: '7',
            key: 'releasenotes/desktop-modeler/7',
            index: 0,
            nodes: {
              '7.0': {
                label: '7.0',
                key: 'releasenotes/desktop-modeler/7.0',
                index: 0,
              },
            },
          },
        },
      },
    },
  },
};
const mockDataWithSlashes = [
  {
    key: 'item1/a',
    label: 'Item 1 (a)',
  },
  {
    key: 'item2/a',
    label: 'Item 2 (a)',
    nodes: [
      {
        key: 'b',
        label: 'Item 2 ab',
        nodes:[
          {
            key: 'b/1',
            label: 'Item 2 ab:1'
          },
          {
            key: 'b/2',
            label: 'Item 2 ab:2'
          },
        ],
      },
      {
        key: 'c',
        label: 'Item 2 ac',
      },
    ],
  },
];

describe('TreeMenu', () => {
  it('should render the level-1 nodes by default', () => {
    const wrapper = mount(<TreeMenu data={mockData} />);

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('li').length).toEqual(2);
  });
  it('should open specified nodes', () => {
    const wrapper = mount(
      <TreeMenu
        data={mockData}
        openNodes={['releasenotes', 'releasenotes/desktop-modeler']}
      />
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('li').length).toEqual(4);
  });
  it('should highlight the active node', () => {
    const activeKey = 'releasenotes/desktop-modeler/7';
    const wrapper = mount(
      <TreeMenu
        data={mockData}
        activeKey={activeKey}
        openNodes={['releasenotes', 'releasenotes/desktop-modeler']}
      />
    );
    const highlightedElement = wrapper
      .findWhere(node => node.key() === activeKey)
      .childAt(0)
      .get(0);

    expect(wrapper).toMatchSnapshot();
    expect(highlightedElement.props.className).toContain('rstm-tree-item--active');
  });
  it('should work with an alternative key separator', () => {
    const wrapper = mount(
      <TreeMenu
        data={mockDataWithSlashes}
        openNodes={['item2/a', 'item2/a;b']}
        keySeparator=";"
      />
    );
    const targetNodeWithAlternativeKey = wrapper.findWhere(node => node.key() === 'item2/a;b').get(0);
    expect(targetNodeWithAlternativeKey).not.toBeUndefined();
    const targetNodeWithAlternativeKey2 = wrapper.findWhere(node => node.key() === 'item2/a;b;b/2').get(0);
    expect(targetNodeWithAlternativeKey2).not.toBeUndefined();
    const targetNodeWithDefaultKey = wrapper.findWhere(node => node.key() === 'item2/a/b').get(0);
    expect(targetNodeWithDefaultKey).toBeUndefined();
  });
  it('should trigger onClickItem when a node is clicked', () => {
    const mockOnClickItem = jest.fn();
    const wrapper = shallow(
      <TreeMenu data={mockData} onClickItem={mockOnClickItem} />
    );

    const targetNode = wrapper.findWhere(node => node.key() === 'releasenotes');
    targetNode.simulate('click');
    expect(mockOnClickItem.mock.calls.length).toEqual(1);
    expect(mockOnClickItem).toHaveBeenCalledWith({
      hasNodes: true,
      index: 0,
      isOpen: false,
      key: 'releasenotes',
      keySeparator: '/',
      label: 'Release Notes',
      level: 0,
      openNodes: [],
      parent: '',
      searchTerm: '',
    });
  });
});
