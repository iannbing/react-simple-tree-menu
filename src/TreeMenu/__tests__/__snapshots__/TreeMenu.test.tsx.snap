// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeMenu should highlight the active node 1`] = `
<TreeMenu
  activeKey="releasenotes/desktop-modeler/7"
  cacheSearch={true}
  data={
    Object {
      "atd": Object {
        "index": 1,
        "key": "ats",
        "label": "ATS Guide",
      },
      "releasenotes": Object {
        "index": 0,
        "key": "releasenotes",
        "label": "Release Notes",
        "nodes": Object {
          "desktop-modeler": Object {
            "index": 0,
            "key": "releasenotes/desktop-modeler",
            "label": "Desktop Modeler",
            "nodes": Object {
              "7": Object {
                "index": 0,
                "key": "releasenotes/desktop-modeler/7",
                "label": "7",
                "nodes": Object {
                  "7.0": Object {
                    "index": 0,
                    "key": "releasenotes/desktop-modeler/7.0",
                    "label": "7.0",
                  },
                },
              },
            },
          },
        },
      },
    }
  }
  debounceTime={125}
  disableKeyboard={false}
  hasSearch={true}
  keySeparator="/"
  onClickItem={[Function]}
  openNodes={
    Array [
      "releasenotes",
      "releasenotes/desktop-modeler",
    ]
  }
  resetOpenNodesOnDataUpdate={false}
>
  <KeyDown
    down={[Function]}
    enter={[Function]}
    left={[Function]}
    right={[Function]}
    up={[Function]}
  >
    <div
      onKeyDown={[Function]}
      tabIndex={0}
    >
      <input
        aria-label="Type and search"
        className="rstm-search"
        onChange={[Function]}
        placeholder="Type and search"
        type="search"
      />
      <ul
        className="rstm-tree-item-group"
      >
        <Component
          active={false}
          focused={false}
          hasNodes={true}
          index={0}
          isOpen={true}
          key="releasenotes"
          keySeparator="/"
          label="Release Notes"
          level={0}
          onClick={[Function]}
          openNodes={
            Array [
              "releasenotes",
              "releasenotes/desktop-modeler",
            ]
          }
          parent=""
          searchTerm=""
          toggleNode={[Function]}
        >
          <li
            aria-pressed={false}
            className="rstm-tree-item rstm-tree-item-level0"
            onClick={[Function]}
            role="button"
            style={
              Object {
                "paddingLeft": "0.75rem",
              }
            }
          >
            <div
              className="rstm-toggle-icon"
              onClick={[Function]}
            >
              <ToggleIcon
                closedIcon="+"
                on={true}
                openedIcon="-"
              >
                <div
                  aria-label="Toggle"
                  className="rstm-toggle-icon-symbol"
                  role="img"
                >
                  -
                </div>
              </ToggleIcon>
            </div>
            Release Notes
          </li>
        </Component>
        <Component
          active={false}
          focused={false}
          hasNodes={true}
          index={0}
          isOpen={true}
          key="releasenotes/desktop-modeler"
          keySeparator="/"
          label="Desktop Modeler"
          level={1}
          onClick={[Function]}
          openNodes={
            Array [
              "releasenotes",
              "releasenotes/desktop-modeler",
            ]
          }
          parent="releasenotes"
          searchTerm=""
          toggleNode={[Function]}
        >
          <li
            aria-pressed={false}
            className="rstm-tree-item rstm-tree-item-level1"
            onClick={[Function]}
            role="button"
            style={
              Object {
                "paddingLeft": "2.5rem",
              }
            }
          >
            <div
              className="rstm-toggle-icon"
              onClick={[Function]}
            >
              <ToggleIcon
                closedIcon="+"
                on={true}
                openedIcon="-"
              >
                <div
                  aria-label="Toggle"
                  className="rstm-toggle-icon-symbol"
                  role="img"
                >
                  -
                </div>
              </ToggleIcon>
            </div>
            Desktop Modeler
          </li>
        </Component>
        <Component
          active={true}
          focused={false}
          hasNodes={true}
          index={0}
          isOpen={false}
          key="releasenotes/desktop-modeler/7"
          keySeparator="/"
          label="7"
          level={2}
          onClick={[Function]}
          openNodes={
            Array [
              "releasenotes",
              "releasenotes/desktop-modeler",
            ]
          }
          parent="releasenotes/desktop-modeler"
          searchTerm=""
          toggleNode={[Function]}
        >
          <li
            aria-pressed={true}
            className="rstm-tree-item rstm-tree-item-level2 rstm-tree-item--active"
            onClick={[Function]}
            role="button"
            style={
              Object {
                "paddingLeft": "4.25rem",
              }
            }
          >
            <div
              className="rstm-toggle-icon"
              onClick={[Function]}
            >
              <ToggleIcon
                closedIcon="+"
                on={false}
                openedIcon="-"
              >
                <div
                  aria-label="Toggle"
                  className="rstm-toggle-icon-symbol"
                  role="img"
                >
                  +
                </div>
              </ToggleIcon>
            </div>
            7
          </li>
        </Component>
        <Component
          active={false}
          focused={false}
          hasNodes={false}
          index={1}
          isOpen={false}
          key="atd"
          keySeparator="/"
          label="ATS Guide"
          level={0}
          onClick={[Function]}
          openNodes={
            Array [
              "releasenotes",
              "releasenotes/desktop-modeler",
            ]
          }
          parent=""
          searchTerm=""
        >
          <li
            aria-pressed={false}
            className="rstm-tree-item rstm-tree-item-level0"
            onClick={[Function]}
            role="button"
            style={
              Object {
                "paddingLeft": "2.75rem",
              }
            }
          >
            ATS Guide
          </li>
        </Component>
      </ul>
    </div>
  </KeyDown>
</TreeMenu>
`;

exports[`TreeMenu should open specified nodes 1`] = `
<TreeMenu
  cacheSearch={true}
  data={
    Object {
      "atd": Object {
        "index": 1,
        "key": "ats",
        "label": "ATS Guide",
      },
      "releasenotes": Object {
        "index": 0,
        "key": "releasenotes",
        "label": "Release Notes",
        "nodes": Object {
          "desktop-modeler": Object {
            "index": 0,
            "key": "releasenotes/desktop-modeler",
            "label": "Desktop Modeler",
            "nodes": Object {
              "7": Object {
                "index": 0,
                "key": "releasenotes/desktop-modeler/7",
                "label": "7",
                "nodes": Object {
                  "7.0": Object {
                    "index": 0,
                    "key": "releasenotes/desktop-modeler/7.0",
                    "label": "7.0",
                  },
                },
              },
            },
          },
        },
      },
    }
  }
  debounceTime={125}
  disableKeyboard={false}
  hasSearch={true}
  keySeparator="/"
  onClickItem={[Function]}
  openNodes={
    Array [
      "releasenotes",
      "releasenotes/desktop-modeler",
    ]
  }
  resetOpenNodesOnDataUpdate={false}
>
  <KeyDown
    down={[Function]}
    enter={[Function]}
    left={[Function]}
    right={[Function]}
    up={[Function]}
  >
    <div
      onKeyDown={[Function]}
      tabIndex={0}
    >
      <input
        aria-label="Type and search"
        className="rstm-search"
        onChange={[Function]}
        placeholder="Type and search"
        type="search"
      />
      <ul
        className="rstm-tree-item-group"
      >
        <Component
          active={false}
          focused={false}
          hasNodes={true}
          index={0}
          isOpen={true}
          key="releasenotes"
          keySeparator="/"
          label="Release Notes"
          level={0}
          onClick={[Function]}
          openNodes={
            Array [
              "releasenotes",
              "releasenotes/desktop-modeler",
            ]
          }
          parent=""
          searchTerm=""
          toggleNode={[Function]}
        >
          <li
            aria-pressed={false}
            className="rstm-tree-item rstm-tree-item-level0"
            onClick={[Function]}
            role="button"
            style={
              Object {
                "paddingLeft": "0.75rem",
              }
            }
          >
            <div
              className="rstm-toggle-icon"
              onClick={[Function]}
            >
              <ToggleIcon
                closedIcon="+"
                on={true}
                openedIcon="-"
              >
                <div
                  aria-label="Toggle"
                  className="rstm-toggle-icon-symbol"
                  role="img"
                >
                  -
                </div>
              </ToggleIcon>
            </div>
            Release Notes
          </li>
        </Component>
        <Component
          active={false}
          focused={false}
          hasNodes={true}
          index={0}
          isOpen={true}
          key="releasenotes/desktop-modeler"
          keySeparator="/"
          label="Desktop Modeler"
          level={1}
          onClick={[Function]}
          openNodes={
            Array [
              "releasenotes",
              "releasenotes/desktop-modeler",
            ]
          }
          parent="releasenotes"
          searchTerm=""
          toggleNode={[Function]}
        >
          <li
            aria-pressed={false}
            className="rstm-tree-item rstm-tree-item-level1"
            onClick={[Function]}
            role="button"
            style={
              Object {
                "paddingLeft": "2.5rem",
              }
            }
          >
            <div
              className="rstm-toggle-icon"
              onClick={[Function]}
            >
              <ToggleIcon
                closedIcon="+"
                on={true}
                openedIcon="-"
              >
                <div
                  aria-label="Toggle"
                  className="rstm-toggle-icon-symbol"
                  role="img"
                >
                  -
                </div>
              </ToggleIcon>
            </div>
            Desktop Modeler
          </li>
        </Component>
        <Component
          active={false}
          focused={false}
          hasNodes={true}
          index={0}
          isOpen={false}
          key="releasenotes/desktop-modeler/7"
          keySeparator="/"
          label="7"
          level={2}
          onClick={[Function]}
          openNodes={
            Array [
              "releasenotes",
              "releasenotes/desktop-modeler",
            ]
          }
          parent="releasenotes/desktop-modeler"
          searchTerm=""
          toggleNode={[Function]}
        >
          <li
            aria-pressed={false}
            className="rstm-tree-item rstm-tree-item-level2"
            onClick={[Function]}
            role="button"
            style={
              Object {
                "paddingLeft": "4.25rem",
              }
            }
          >
            <div
              className="rstm-toggle-icon"
              onClick={[Function]}
            >
              <ToggleIcon
                closedIcon="+"
                on={false}
                openedIcon="-"
              >
                <div
                  aria-label="Toggle"
                  className="rstm-toggle-icon-symbol"
                  role="img"
                >
                  +
                </div>
              </ToggleIcon>
            </div>
            7
          </li>
        </Component>
        <Component
          active={false}
          focused={false}
          hasNodes={false}
          index={1}
          isOpen={false}
          key="atd"
          keySeparator="/"
          label="ATS Guide"
          level={0}
          onClick={[Function]}
          openNodes={
            Array [
              "releasenotes",
              "releasenotes/desktop-modeler",
            ]
          }
          parent=""
          searchTerm=""
        >
          <li
            aria-pressed={false}
            className="rstm-tree-item rstm-tree-item-level0"
            onClick={[Function]}
            role="button"
            style={
              Object {
                "paddingLeft": "2.75rem",
              }
            }
          >
            ATS Guide
          </li>
        </Component>
      </ul>
    </div>
  </KeyDown>
</TreeMenu>
`;

exports[`TreeMenu should render the level-1 nodes by default 1`] = `
<TreeMenu
  cacheSearch={true}
  data={
    Object {
      "atd": Object {
        "index": 1,
        "key": "ats",
        "label": "ATS Guide",
      },
      "releasenotes": Object {
        "index": 0,
        "key": "releasenotes",
        "label": "Release Notes",
        "nodes": Object {
          "desktop-modeler": Object {
            "index": 0,
            "key": "releasenotes/desktop-modeler",
            "label": "Desktop Modeler",
            "nodes": Object {
              "7": Object {
                "index": 0,
                "key": "releasenotes/desktop-modeler/7",
                "label": "7",
                "nodes": Object {
                  "7.0": Object {
                    "index": 0,
                    "key": "releasenotes/desktop-modeler/7.0",
                    "label": "7.0",
                  },
                },
              },
            },
          },
        },
      },
    }
  }
  debounceTime={125}
  disableKeyboard={false}
  hasSearch={true}
  keySeparator="/"
  onClickItem={[Function]}
  resetOpenNodesOnDataUpdate={false}
>
  <KeyDown
    down={[Function]}
    enter={[Function]}
    left={[Function]}
    right={[Function]}
    up={[Function]}
  >
    <div
      onKeyDown={[Function]}
      tabIndex={0}
    >
      <input
        aria-label="Type and search"
        className="rstm-search"
        onChange={[Function]}
        placeholder="Type and search"
        type="search"
      />
      <ul
        className="rstm-tree-item-group"
      >
        <Component
          active={false}
          focused={false}
          hasNodes={true}
          index={0}
          isOpen={false}
          key="releasenotes"
          keySeparator="/"
          label="Release Notes"
          level={0}
          onClick={[Function]}
          openNodes={Array []}
          parent=""
          searchTerm=""
          toggleNode={[Function]}
        >
          <li
            aria-pressed={false}
            className="rstm-tree-item rstm-tree-item-level0"
            onClick={[Function]}
            role="button"
            style={
              Object {
                "paddingLeft": "0.75rem",
              }
            }
          >
            <div
              className="rstm-toggle-icon"
              onClick={[Function]}
            >
              <ToggleIcon
                closedIcon="+"
                on={false}
                openedIcon="-"
              >
                <div
                  aria-label="Toggle"
                  className="rstm-toggle-icon-symbol"
                  role="img"
                >
                  +
                </div>
              </ToggleIcon>
            </div>
            Release Notes
          </li>
        </Component>
        <Component
          active={false}
          focused={false}
          hasNodes={false}
          index={1}
          isOpen={false}
          key="atd"
          keySeparator="/"
          label="ATS Guide"
          level={0}
          onClick={[Function]}
          openNodes={Array []}
          parent=""
          searchTerm=""
        >
          <li
            aria-pressed={false}
            className="rstm-tree-item rstm-tree-item-level0"
            onClick={[Function]}
            role="button"
            style={
              Object {
                "paddingLeft": "2.75rem",
              }
            }
          >
            ATS Guide
          </li>
        </Component>
      </ul>
    </div>
  </KeyDown>
</TreeMenu>
`;
